src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1114,6 +1121,219 @@ bool PhaseIdealLoop::create_loop_nest(IdealLoopTree* loop, Node_List &old_new) {
+    if (deopt_reason == Deoptimization::Reason_predicate ||
+        deopt_reason == Deoptimization::Reason_profile_predicate) {
+      _has_hoisted_check_parse_predicates = true;
+    }
+
+    _clone_predicate_to_loop.clone_parse_predicate(parse_predicate, true);
+    parse_predicate.kill(_phase->igvn());
+  }
+  void visit(const TemplateAssertionPredicate& template_assertion_predicate) override {
+    if (!_has_hoisted_check_parse_predicates) {
+      // Only process if we are in the correct Predicate Block.
+      return;
+    }
+
+    _clone_predicate_to_loop.clone_template_assertion_predicate(template_assertion_predicate);
+    template_assertion_predicate.kill(_phase->igvn());
+  }
+};
+
+// If bounds are known is the loop doesn't need an outer loop or profile data indicates it runs for less than
+// ShortLoopIter, don't create the outer loop
+bool PhaseIdealLoop::short_running_loop(IdealLoopTree* loop, jint stride_con, const Node_List &range_checks, uint iters_limit) {
+  if (!ShortRunningLongLoop) {
+    return false;
+  }
+  Node* x = loop->_head;
+  BaseCountedLoopNode* head = x->as_BaseCountedLoop();
+  BasicType bt = head->bt();
+  Node* entry_control = head->skip_strip_mined()->in(LoopNode::EntryControl);
+
+  loop->compute_trip_count(this, bt);
+  // Loop must run for no more than iter_limits as it guarantees no overflow of scale * iv in long range checks.
+  bool known_short_running_loop = head->trip_count() <= iters_limit / ABS(stride_con);
+  bool profile_short_running_loop = false;
+  if (!known_short_running_loop) {
+    loop->compute_profile_trip_cnt(this);
+    if (StressShortRunningLongLoop) {
+     profile_short_running_loop = true;
+    } else {
+      profile_short_running_loop = !head->is_profile_trip_failed() && head->profile_trip_cnt() < iters_limit / ABS(stride_con);
+    }
+  }
+
+  if (!known_short_running_loop && !profile_short_running_loop) {
+    return false;
+  }
+
+  Node* limit = head->limit();
+  Node* init = head->init_trip();
+
+  Node* new_limit;
+  if (stride_con > 0) {
+    new_limit = SubNode::make(limit, init, bt);
+  } else {
+    new_limit = SubNode::make(init, limit, bt);
+  }
+  register_new_node(new_limit, entry_control);
+
+  PhiNode* phi = head->phi()->as_Phi();
+  const Type* new_phi_t = TypeInt::INT;
+  if (profile_short_running_loop) {
+    // Add a short_limit predicate. It's the last predicate when coming from the loop because a cast that's control
+    // dependent on the short_limit predicate is added to narrow the limit and future predicates may be dependent on the
+    // new limit (so have to be between the loop and short_limit predicate). The current limit could, itself, be
+    // dependent on an existing predicate. Clone parse predicates below existing predicates to get proper ordering of
+    // predicates when coming from the loop: future predicates, short_limit predicate, existing predicates.
```

Comment:
Maybe be more explicit:
```suggestion
    // predicates when walking from the loop up: future predicates, short_limit predicate, existing predicates.
```

A visualization might also help here to quickly grasp the structure:
```
//        Existing Hoisted 
//        Check Predicates  
//              |         
//     New Short Running Long        
//         Loop Predicate
//              |
//   Cloned Parse Predicates and 
//  Template Assertion Predicates        
//              |
//             Loop
```